### PR TITLE
fixes #1431 - On app deletion only remove app definition from repo when ...

### LINKED ISSUE #1431: When deleting an app, don't remove app definition before all task are killed
This way the decreasing task count will be visible in the API / UI

### PR DESCRIPTION
**Pull request by drexin**
...all tasks are killed

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/upgrade/AppStopActor.scala [hunk 1]
```diff
@@ -24,6 +24,6 @@ class AppStopActor(
 
   def initializeStop(): Unit = {
     eventBus.subscribe(self, classOf[MesosStatusUpdateEvent])
-    scheduler.stopApp(driver, app)
+    for (id <- idsToKill) driver.killTask(Protos.TaskID.newBuilder.setValue(id).build())
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/upgrade/AppStopActor.scala]: I guess that you can get rid of the scheduler dependency in AppStopActor then.

File: src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala [hunk 2]
```diff
@@ -339,8 +339,16 @@ class MarathonSchedulerActorTest extends TestKit(ActorSystem("System"))
     val plan = DeploymentPlan("foo", origGroup, targetGroup, List(DeploymentStep(List(StopApplication(app)))), Timestamp.now())
 
     when(tracker.get(app.id)).thenReturn(Set(taskA))
+    when(repo.store(any())).thenReturn(Future.successful(app))
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: would it maybe make sense to check that the instance count got changed to zero?

File: src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala [hunk 3]
```diff
@@ -85,7 +85,7 @@ class DeploymentActor(
           case StartApplication(app, scaleTo) => storeAndThen(app) { startApp(app, scaleTo) }
           case ScaleApplication(app, scaleTo) => storeAndThen(app) { scaleApp(app, scaleTo) }
           case RestartApplication(app)        => storeAndThen(app) { restartApp(app) }
-          case StopApplication(app)           => stopApp(app)
+          case StopApplication(app)           => storeAndThen(app.copy(instances = 0)) { stopApp(app) }
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala]: This change doesn't have any automatic tests associated with it. If I revert the change, all tests still work.

**Comment by kolloch** [comment]
LGTM except comments. I would appreciate more tests.

File: src/main/scala/mesosphere/marathon/upgrade/AppStopActor.scala (hunk 1 above)
Comment (drexin) [inline src/main/scala/mesosphere/marathon/upgrade/AppStopActor.scala]: Good catch.

File: src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala (hunk 2 above)
Comment (drexin) [inline src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: That test only checks if the rate limiter gets reset. I don't think we should check the instance count here.

Comment (drexin) [inline src/test/scala/mesosphere/marathon/MarathonSchedulerActorTest.scala]: But I will add a test for that.

File: src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala (hunk 3 above)
Comment (drexin) [inline src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala]: Will add a test.

**Comment by drexin** [comment]
@kolloch Updated PR.

**Comment by kolloch** [comment]
:)
